### PR TITLE
fix(cli): remove usage of deprecated npm env var from arg parser

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This removes some code from our argument parsing implementation which
would look for an environment variable (`npm_config_argv`) set by npm
and, if present, merge CLI flags found their into the ones already
present on `process.argv`.

We want to remove this for two reasons:

- `npm_config_argv` is deprecated in `npm`, so in newer versions of the
  package manager it is no longer set and our code referencing it is
  effectively doing nothing
- `yarn` v1.x still sets it (presumably for compatibility with `npm`
  that hasn't been removed yet, which causes an inconsistency between
  `npm` and `yarn` where certain `package.json` scripts will run without
  issue in `npm` but fail in `yarn` (see #3482)

Accordingly, this commit deletes the offending function from
`src/cli/parse-flags.ts` and makes a few related changes at call sites,
etc which are necessary due to that change.

This commit also refactors the spec file for `parse-flags.ts` to remove
redundant tests. Because all of the supported CLI arguments (i.e. those
added to `knownArgs`) are defined in `ReadonlyArray<string>` variables
we can do a lot of exhaustive testing of all the arguments, set in the
various possible permutations, etc, so we don't need to define a bunch
of individual tests. Accordingly, the test file is refactored to remove
redundant tests, add a few more test cases for the exhaustive tests,
organize things a bit more with `describe` blocks, and ensure that we
have good tests around all off the 'edge-case-ey' things.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

When parsing CLI args (in `parse_flags.ts`) Stencil will look at `process.argv` and _also_ at an environment variable called `npm_config_argv`. This [was deprecated](https://github.com/npm/rfcs/blob/main/implemented/0021-reduce-lifecycle-script-environment.md) in `npm` but is still set by (at least) yarn v1.x. This causes a discrepancy between yarn and npm in what CLI arguments will be parsed by Stencil, and, in particular, this can lead to a situation where different arguments are passed to Jest (see #3482 for some details).

## What is the new behavior?

This removes the code that accesses `npm_config_argv` in favor of just depending on `process.argv`.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No\*

\* This does in a sense change the ways that CLI args can be passed to Stencil. It's possible that users of older versions of npm who depend on the (undocumented) behavior of passing through CLI arguments to npm will experience some disruption if they upgrade. However, given that this code _does nothing_ on newer versions of npm and that hasn't led to any disruption, any impact is likely to be minimal.

## Testing

There are some changes and enhancements to the unit tests for CLI argument parsing module. Additionally, I built and `npm pack`-ed these changes in a few example projects to make sure that various ways of running stencil are fully supported, e.g. `npx stencil test --e2e`, from a package.json script, etc.

Additionally, these things should all work the same between yarn and npm!

### CODE REVIEW

File: src/cli/test/parse-flags.spec.ts [hunk 1]
```diff
@@ -30,556 +20,216 @@ describe('parseFlags', () => {
...
 
-    const flags = parseFlags(args, sys);
+    const flags = parseFlags(args);
     expect(flags.task).toBe('serve');
     expect(flags.address).toBe('127.0.0.1');
     expect(flags.port).toBe(4444);
     expect(flags.knownArgs).toEqual(['--address', '127.0.0.1', '--port', '4444']);
   });
 
-  it('should use cli args first, then npm cmds', () => {
```
Comment (alicewriteswrongs) [inline src/cli/test/parse-flags.spec.ts:49]: this test (and the test below) is for the now-removed usage of the deprecated `npm_config_argv`, and is deleted for that reason

File: src/cli/test/parse-flags.spec.ts [hunk 2]
```diff
@@ -30,556 +20,216 @@ describe('parseFlags', () => {
...
   });
 
   it('should parse no task', () => {
-    args[0] = '--flag';
-    const flags = parseFlags(args, sys);
+    const flags = parseFlags(['--flag']);
     expect(flags.task).toBe(null);
   });
 
-  it('should parse build flag to true', () => {
```
Comment (alicewriteswrongs) [inline src/cli/test/parse-flags.spec.ts:119]: this test (and many others as well) is essentially completely redundant given our exhaustive testing of all boolean flags with `describe.each(BOOLEAN_CLI_ARGS`, below. While I was in here I wanted to remove them so that this file can be smaller, more focused, and easier to read. We have had a few bugs and edge cases crop up recently related to arg parsing, and since we're maintaining our own library for doing this I want to try to make sure this file is easy to read and understand, has exhaustive tests of all the supported arguments, and so on.

File: src/cli/test/parse-flags.spec.ts [hunk 3]
```diff
@@ -30,556 +20,216 @@ describe('parseFlags', () => {
...
+    });
+
+    it(`should parse "--${toDashCase(cliArg)} value"`, () => {
+      const flags = parseFlags([`--${toDashCase(cliArg)}`, 'path/to/file.js']);
+      expect(flags.knownArgs).toEqual([`--${toDashCase(cliArg)}`, 'path/to/file.js']);
+      expect(flags.unknownArgs).toEqual([]);
+      expect(flags[cliArg]).toBe('path/to/file.js');
+    });
+
+    it(`should parse "--${toDashCase(cliArg)}=value"`, () => {
```
Comment (alicewriteswrongs) [inline src/cli/test/parse-flags.spec.ts:106]: I added two more tests for the string arguments here, basically testing that `--foo-bar baz` and `--foo-bar=baz` will be parsed correctly

File: src/cli/test/parse-flags.spec.ts [hunk 4]
```diff
@@ -30,556 +20,216 @@ describe('parseFlags', () => {
...
-    expect(flags.screenshot).toBe(true);
-  });
-
-  it('should parse --screenshot-connector scripts/connector.js', () => {
-    args[0] = '--screenshot-connector';
-    args[1] = 'scripts/connector.js';
-    const flags = parseFlags(args, sys);
-    expect(flags.screenshotConnector).toBe('scripts/connector.js');
-  });
+  describe.each<LogLevel>(['info', 'warn', 'error', 'debug'])('logLevel %s', (level) => {
```
Comment (alicewriteswrongs) [inline src/cli/test/parse-flags.spec.ts:127]: I pulled the `LogLevel` specs together here basically just for clarity and organization

**Review by alicewriteswrongs — COMMENTED** [review]
pointing out a few things

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
Looks good! testing is much more focused now 🎯 I have one concern about a breaking change, but I think there's a pretty good middle ground/path forward here

File: src/cli/parse-flags.ts [hunk 5]
```diff
@@ -344,19 +332,3 @@ const isLogLevel = (maybeLogLevel: string): maybeLogLevel is LogLevel =>
   //
   // see microsoft/TypeScript#31018 for some discussion of this
   LOG_LEVELS.includes(maybeLogLevel as any);
-
-const getNpmConfigEnvArgs = (sys: CompilerSystem) => {
-  // process.env.npm_config_argv
-  // {"remain":["4444"],"cooked":["run","serve","--port","4444"],"original":["run","serve","--port","4444"]}
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/cli/parse-flags.ts:350]: I didn't see anything explaining it...it was somehow a bit difficult to find information about exactly what could end up in this variable!

File: src/cli/parse-flags.ts [hunk 6]
```diff
@@ -19,11 +19,10 @@ import {
 /**
  * Parse command line arguments into a structured `ConfigFlags` object
  *
- * @param args an array of config flags
- * @param sys an optional compiler system
+ * @param args an array of CLI flags
  * @returns a structured ConfigFlags object
  */
-export const parseFlags = (args: string[], sys?: CompilerSystem): ConfigFlags => {
+export const parseFlags = (args: string[]): ConfigFlags => {
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/cli/parse-flags.ts:25]: Ok yeah, I think that makes sense to me. I'll restore the param (w/ a `_`) and add a `TODO` Jira ticket for removing that for v3

(probably will put up a PR against the v3 branch right away as well, just so we don't forget!)

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/cli/parse-flags.ts:25]: filed `STENCIL-509`

[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply src/cli/parse-flags.ts]: Perfectx3 - ty!

File: src/cli/parse-flags.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply src/cli/parse-flags.ts:350]: Yeah, neither could I.  I suppose we were using `'original'`, whose name implies it's an unmodified version of the args 🤔

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/cli/parse-flags.ts:350]: haha, maybe it should be 'raw' vs 'cooked'? my guess would be something like escaping or un-escaping some control sequences, maybe filtering out certain things, idk

**Comment by rwaskiewicz** [comment]
@alicewriteswrongs 

So, GH is having an interesting day in that I can't upload a screenshot ATM.  I tested arg parsing on both this branch and `main` using npm 6 + node 12, as well as npm 8 + node 16.  To get a cheap look at what got parsed (and easily diff them), I applied the following patch:

```diff
diff --git a/src/cli/parse-flags.ts b/src/cli/parse-flags.ts
index d4a14c075..7f1c71155 100644
--- a/src/cli/parse-flags.ts
+++ b/src/cli/parse-flags.ts
@@ -47,6 +47,7 @@ export const parseFlags = (args: string[]): ConfigFlags => {
   // string `"--bizBoz"` will be added.
   flags.unknownArgs = flags.args.filter((arg: string) => !flags.knownArgs.includes(parseEqualsArg(arg)[0]));
 
+  console.log(JSON.stringify(flags, null, 4));
   return flags;
 };
```

For all running scenarios, I ran the following command: `npm run build -- --foo bar run=one --run=two`

The results were exactly the same in Node 16 + npm 8. 

In npm 6 + node 12, I saw one minor difference.  In `main`, the `--` gets parsed as an arg + unknown arg, but does not in this branch:

```
// main output
{
    "task": "build",
    "args": [
        "--docs",
        "--foo",
        "bar",
        "run=one",
        "--run=two",
        "--"
    ],
    "knownArgs": [
        "--docs"
    ],
    "unknownArgs": [
        "--foo",
        "bar",
        "run=one",
        "--run=two",
        "--"
    ],
   ...
```

```
// This branch
  {
    "task": "build",
    "args": [
        "--docs",
        "--foo",
        "bar",
        "run=one",
        "--run=two"
    ],
    "knownArgs": [
        "--docs"
    ],
    "unknownArgs": [
        "--foo",
        "bar",
        "run=one",
        "--run=two"
    ],
  ...
```

Do you think we should be concerned about this change?

File: src/cli/parse-flags.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply src/cli/parse-flags.ts:350]: Lolol - yeah that'd make sense!

**Comment by alicewriteswrongs** [comment]
I'm not really worried about that, for a few reasons. Afaict the main thing that `unknownArgs` ends up being used for is supporting passing arguments that Jest recognized but which we don't know about, see here:

https://github.com/ionic-team/stencil/blob/c0aeac1176c474f4e0935e40f3944433f5c24860/src/testing/jest/jest-config.ts#L41-L76

so when we build up the jest config from CLI args we use both `unknownArgs` and `knownArgs`. This means that even if our manifest of Jest-supported arguments is inaccurate, there's a version mismatch, whatever it might be, then an argument passed on the CLI will still make its way through to Jest and can affect how it runs by setting some boolean flag, sending output to some file, etc.

Anywhoo - in my testing it seems that if `--` is present like that it causes issues with Jest arguments similar to the issue we're trying to fix. To test this out I installed `@stencil/core@2.17.1` in a test component alongisde `npm@6.14.17`, a setup which will read from a populated `npm_config_argv` variable when parsing the CLI args. Here's the result:

Screen Shot 2022-07-22 at 2 08 05 PM

(for the above my `test` command is `stencil test --spec`)

The same thing happens with a recent 1.x version of yarn:

Screen Shot 2022-07-22 at 2 10 29 PM


So I _think_ `--` is interpreted by jest as the string it expects before a pattern to match tests against, so pulling the `--` out of the `npm_config_argv` like this creates what I would call a bug.

Does that all make sense? I feel like my head is full of details about this and I'm not sure I'm explaining it the best 😅

**Comment by rwaskiewicz** [comment]
Yep - SGTM. Thanks for writing that out!

**Review by rwaskiewicz — APPROVED** [review]
